universe/Ship.cpp

```diff
@@ -464,7 +469,11 @@ namespace {
             ShipPartClass part_class = part->Class();
 
             // get the attack power for each weapon part.
-            if (part_class == ShipPartClass::PC_DIRECT_WEAPON) {
+            if (part->TotalShipDamage()) { // FIXME TotalFighterDamage
+                // there is a special computation. overrides the default one
+                double estimation = part->TotalShipDamage()->Eval(context);
```

Comment:
@geoffthemedio here I use the context for the evaluation of the totalShipDamage valueref. But it seems it does not work - I try to access the ship inside the context via Source, but it always returns zero (e.g. Source.Structure)

Reply:
should be able to use `shared_from_this()`